plugins/node/opentelemetry-instrumentation-pulsar-client/src/proxies/consumer.ts

```diff
@@ -0,0 +1,211 @@
+    this.closePreviousSpan();
+    return this.consumer.acknowledgeCumulative(message);
+  }
+
+  acknowledgeCumulativeId(messageId: Pulsar.MessageId): Promise<null> {
+    this.closePreviousSpan();
+    return this.consumer.acknowledgeCumulativeId(messageId);
+  }
+
+  isConnected(): boolean {
+    return this.consumer.isConnected();
+  }
+
+  close(): Promise<null> {
+    this.closePreviousSpan();
+    return this.consumer.close();
+  }
+
+  unsubscribe(): Promise<null> {
+    this.closePreviousSpan();
+    return this.consumer.unsubscribe();
+  }
+}
+
+function trackBeforeAndAfterConsume(tracer: Tracer, spanName: string, extraAttributes: Attributes) {
+  tracer.startActiveSpan(spanName, {attributes: extraAttributes}, span => {
+    span.end()
+  });
+}
+
+function getAttributesFromMessage(message: Pulsar.Message) {
+  return {
+    [SemanticAttributes.MESSAGING_DESTINATION]: message.getTopicName(),
+    [SemanticAttributes.MESSAGING_MESSAGE_ID]: message.getMessageId().toString(),
+  };
+}
+
+function extractSpanFromMessage(
+  tracer: api.Tracer,
+  instrumentationConfig: PulsarInstrumentationConfig,
+  moduleVersion: string | undefined,
+  config: Pulsar.ConsumerConfig,
+  message: Pulsar.Message
+) {
+  const remoteContext = api.propagation.extract(
+    api.context.active(),
+    message.getProperties()
+  );
+
+  const extraAttributes = getAttributesFromMessage(message);
+  if (instrumentationConfig.trackBeforeAndAfterConsume) {
+    trackBeforeAndAfterConsume(tracer, spanNames.beforeConsume, extraAttributes);
+  }
+
+  const span = tracer.startSpan(
+    'receive',
+    {
+      kind: api.SpanKind.CONSUMER,
+      attributes: {
+        'pulsar.version': moduleVersion,
+        ...extraAttributes,
+        ...Instrumentation.COMMON_ATTRIBUTES,
+      },
+    },
+    remoteContext
+  );
+
+  api.trace.setSpan(remoteContext, span);
+
+  return span;
+}
+
+export function wrappedListener(
+  tracer: Tracer,
+  instrumentationConfig: PulsarInstrumentationConfig,
+  moduleVersion: string | undefined,
+  config: ConsumerConfig,
+  listener: ConsumerListener
+): ConsumerListener {
+  return async (message: Pulsar.Message, consumer: Pulsar.Consumer) => {
+    const span = extractSpanFromMessage(tracer, instrumentationConfig, moduleVersion, config, message);
+    try {
+      await callback(listener, message, consumer);
+    } catch (error) {
+      span.recordException(error);
+      span.setStatus({code: SpanStatusCode.ERROR});
+      throw error;
+    } finally {
+      span.end();
+      if (instrumentationConfig.trackBeforeAndAfterConsume) {
+        trackBeforeAndAfterConsume(tracer, spanNames.afterConsume, getAttributesFromMessage(message));
+      }
+    }
+  };
+}
+
+async function callback(
+  original: ConsumerListener,
+  message: Pulsar.Message,
+  consumer: Pulsar.Consumer
+) {
+  // Deal with both sync and asynchronous functions
+  const rv = original(message, consumer);
```

Comment:
I think that if you invoke `original` like this, it will not have the proper `this` value of the original call. This can possibly introduce a change for users after installing the instrumentation. 

If you search the repo for `original.apply` you can see examples of how the `this` parameter should be propagated to the patched function. Sometimes `original.call` is also used.